File: Orientada a Objeto/Caio/personagem.py
class Personagem:
    """Representa um personagem em um jogo de RPG.

    Gerencia o estado do personagem usando Propriedades Python
    para garantir que as regras de negócio (limites de vida/mana)
    sejam sempre aplicadas de forma centralizada.
    """

    VIDA_MAXIMA = 100
    VIDA_MINIMA = 0
    MANA_MAXIMA = 50
    MANA_MINIMA = 0

    def __init__(self, nome: str):
        """Inicializa um novo Personagem."""
        self.nome = nome
        self.pontosVida = self.VIDA_MAXIMA
        self.pontosMana = self.MANA_MAXIMA

    # --- Propriedade NOME ---
    @property
    def nome(self) -> str:
        return self.__nome

    @nome.setter
    def nome(self, novo_nome: str):
        if novo_nome and novo_nome.strip():
            self.__nome = novo_nome.strip()
        else:
            raise ValueError("O nome não pode ser vazio")

    # --- Propriedade PONTOS DE VIDA ---
    @property
    def pontosVida(self) -> int:
        return self.__pontosVida

    @pontosVida.setter
    def pontosVida(self, novo_valor: int):
        self.__pontosVida = max(self.VIDA_MINIMA,
                                min(self.VIDA_MAXIMA, novo_valor))

    # --- Propriedade PONTOS DE MANA ---
    @property
    def pontosMana(self) -> int:
        return self.__pontosMana

    @pontosMana.setter
    def pontosMana(self, novo_valor: int):
        self.__pontosMana = max(self.MANA_MINIMA,
                                min(self.MANA_MAXIMA, novo_valor))

    # --- MÉTODOS DE AÇÃO ---

    def tomarDano(self, dano: int):
        """Aplica dano ao personagem."""
        if dano > 0:
            self.pontosVida = self.pontosVida - dano

    def curar(self, cura: int):
        """Aplica cura ao personagem."""
        if cura > 0:
            self.pontosVida = self.pontosVida + cura

    def usarMagia(self, custoMana: int) -> bool:
        """Tenta usar uma magia, consumindo mana."""
        if custoMana <= 0:
            return False

        if self.pontosMana >= custoMana:
            self.pontosMana = self.pontosMana - custoMana
            print(f"{self.nome} usou magia!")
            return True
        else:
            print(f"{self.nome} tentou usar magia, mas não tem mana!")
            return False

    def estaVivo(self) -> bool:
        """Verifica se o personagem está vivo."""
        return self.pontosVida > self.VIDA_MINIMA

File: Orientada a Objeto/Wallace/Personagem.java
/** Classe que define a estrutura de um Personagem, aplicando Encapsulamento. */
public class Personagem {

    // A. Atributos Privados (Encapsulamento)
    private String nome;
    private int pontosVida;
    private int pontosMana;
    private static final int VIDA_MAXIMA = 100; 

    // B. Métodos Públicos (Comportamentos)

    // Construtor
    public Personagem(String nome) {

        this.nome = nome;
        this.pontosVida = VIDA_MAXIMA;
        this.pontosMana = 50;
        
        System.out.println("Personagem " + nome + " foi criado!"); 
    }

    // Métodos de Ação

    /** Subtrai dano dos pontos de vida. Os PVs não podem ser menores que 0.*/
    public void tomarDano(int dano) {
        if (dano > 0) {
            this.pontosVida -= dano;
            // Regra: Pontos de vida não podem ficar abaixo de 0
            if (this.pontosVida < 0) {
                this.pontosVida = 0;
            }
            System.out.println(this.nome + " tomou " + dano + " de dano. PVs atuais: " + this.pontosVida);
        }
    }


    /** Adiciona cura aos pontos de vida. Os PVs não podem ultrapassar 100. */
    public void curar(int cura) {
        if (cura > 0) {
            this.pontosVida += cura;
            // Regra: Pontos de vida não podem ultrapassar o valor máximo
            if (this.pontosVida > VIDA_MAXIMA) {
                this.pontosVida = VIDA_MAXIMA;
            }
            System.out.println(this.nome + " se curou em " + cura + " pontos. PVs atuais: " + this.pontosVida);
        }
        
    }


    /** Tenta usar uma magia, consumindo mana se houver o suficiente. */
    public boolean usarMagia(int custoMana) {
        if (custoMana <= 0) {
             System.out.println(this.nome + " tentou usar magia com custo 0 ou negativo. Operação cancelada.");
             return false;
        }
        
        // Regra: Verifica se há mana suficiente
        if (this.pontosMana >= custoMana) {
            this.pontosMana -= custoMana;
            System.out.println(this.nome + " usou magia com sucesso! Custo: " + custoMana + ". Mana restante: " + this.pontosMana);
            return true;
        } else {
            System.out.println(this.nome + " falhou ao usar magia. Mana insuficiente.");
            return false;
        }
    }


    // Métodos de Consulta (Getters)

    public String getNome() {
        return nome;
    }

    public int getPontosVida() {
        return pontosVida;
    }

    public int getPontosMana() {
        return pontosMana;
    }

    /** Retorna se o personagem está vivo (pontosVida > 0). */
    public boolean estaVivo() {
        return pontosVida > 0;
    }
}